File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/config/updating/EcoConfigHandler.java
package com.willfp.eco.internal.config.updating;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.PluginDependent;
import com.willfp.eco.core.config.interfaces.LoadableConfig;
import com.willfp.eco.core.config.updating.ConfigHandler;
import com.willfp.eco.core.config.updating.ConfigUpdater;
import com.willfp.eco.internal.config.updating.exceptions.InvalidUpdateMethodException;
import com.willfp.eco.internal.config.yaml.EcoUpdatableYamlConfig;
import org.jetbrains.annotations.NotNull;
import org.reflections.Reflections;
import org.reflections.scanners.MethodAnnotationsScanner;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.List;

public class EcoConfigHandler extends PluginDependent<EcoPlugin> implements ConfigHandler {
    private final List<LoadableConfig> configs = new ArrayList<>();

    private final Reflections reflections = new Reflections(this.getPlugin().getClass().getPackageName(), new MethodAnnotationsScanner());

    public EcoConfigHandler(@NotNull final EcoPlugin plugin) {
        super(plugin);
    }

    @Override
    public void callUpdate() {
        for (Method method : reflections.getMethodsAnnotatedWith(ConfigUpdater.class)) {
            if (!Modifier.isStatic(method.getModifiers())) {
                throw new InvalidUpdateMethodException("Update method must be static.");
            }

            try {
                if (method.getParameterCount() == 0) {
                    method.invoke(null);
                } else if (method.getParameterCount() == 1) {
                    method.invoke(null, this.getPlugin());
                } else {
                    throw new InvalidUpdateMethodException("Update method must have 0 parameters or a plugin parameter.");
                }
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
                throw new InvalidUpdateMethodException("Update method generated an exception.");
            }
        }
    }

    @Override
    public void saveAllConfigs() {
        try {
            for (LoadableConfig config : configs) {
                config.save();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void addConfig(@NotNull final LoadableConfig config) {
        configs.add(config);
    }

    @Override
    public void updateConfigs() {
        for (LoadableConfig config : configs) {
            if (config instanceof EcoUpdatableYamlConfig updatableYamlConfig) {
                updatableYamlConfig.update();
            }
        }
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/config/updating/EcoConfigHandler.kt
package com.willfp.eco.internal.config.updating

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.config.interfaces.LoadableConfig
import com.willfp.eco.core.config.updating.ConfigHandler
import com.willfp.eco.core.config.updating.ConfigUpdater
import com.willfp.eco.internal.config.json.EcoLoadableJSONConfig
import com.willfp.eco.internal.config.json.EcoUpdatableJSONConfig
import com.willfp.eco.internal.config.updating.exceptions.InvalidUpdateMethodException
import com.willfp.eco.internal.config.yaml.EcoLoadableYamlConfig
import com.willfp.eco.internal.config.yaml.EcoUpdatableYamlConfig
import org.reflections.Reflections
import org.reflections.scanners.Scanners
import org.reflections.util.ConfigurationBuilder

class EcoConfigHandler(
    private val plugin: EcoPlugin
) : ConfigHandler {
    private val reflections: Reflections = Reflections(
        ConfigurationBuilder()
            .addClassLoaders(this.plugin::class.java.classLoader)
            .addScanners(Scanners.MethodsAnnotated)
    )

    private val configs = mutableListOf<LoadableConfig>()

    override fun callUpdate() {
        for (method in reflections.getMethodsAnnotatedWith(ConfigUpdater::class.java)) {
            try {
                when (method.parameterCount) {
                    0 -> method.invoke(null)
                    1 -> method.invoke(null, this.plugin)
                    else -> throw InvalidUpdateMethodException("Update method must have 0 parameters or a plugin parameter.")
                }
            } catch (e: ReflectiveOperationException) {
                e.printStackTrace()
                throw InvalidUpdateMethodException("Update method generated an exception")
            }
        }
    }

    override fun saveAllConfigs() {
        for (config in configs) {
            config.save()
        }
    }

    override fun addConfig(config: LoadableConfig) {
        configs.add(config)
    }

    override fun updateConfigs() {
        for (config in configs) {
            when (config) {
                is EcoUpdatableYamlConfig -> config.update()
                is EcoUpdatableJSONConfig -> config.update()
                is EcoLoadableYamlConfig -> config.reloadFromFile()
                is EcoLoadableJSONConfig -> config.reloadFromFile()
            }
        }
    }
}